summary/percentiles.go

```diff
@@ -0,0 +1,183 @@
+		return 0
+	}
+	sort.Sort(self)
+	n := float64(0.9 * (float64(count) + 1))
+	idx, frac := math.Modf(n)
+	index := int(idx)
+	percentile := float64(self[index-1])
+	if index > 1 || index < count {
+		percentile += frac * float64(self[index]-self[index-1])
+	}
+	return uint64(percentile)
+}
+
+type Mean struct {
+	// current count.
+	count uint64
+	// current mean.
+	Mean float64
+}
+
+func (self *Mean) Add(value uint64) {
+	self.count++
+	if self.count == 1 {
+		self.Mean = float64(value)
+		return
+	}
+	c := float64(self.count)
+	v := float64(value)
+	self.Mean = (self.Mean*(c-1) + v) / c
+}
+
+type Resource struct {
+	// list of samples being tracked.
+	samples uint64Slice
+	// average from existing samples.
+	mean Mean
+	// maximum value seen so far in the added samples.
+	max uint64
+}
+
+// Adds a new percentile sample.
+func (self *Resource) Add(p info.Percentiles) {
+	if !p.Present {
+		return
+	}
+	if p.Max > self.max {
+		self.max = p.Max
+	}
+	self.mean.Add(p.Mean)
+	// Selecting 90p of 90p :(
+	self.samples = append(self.samples, p.Ninety)
+}
+
+// Add a single sample. Internally, we convert it to a fake percentile sample.
+func (self *Resource) AddSample(val uint64) {
+	sample := info.Percentiles{
+		Present: true,
+		Mean:    val,
+		Max:     val,
+		Ninety:  val,
+	}
+	self.Add(sample)
+}
+
+// Get max, average, and 90p from existing samples.
+func (self *Resource) GetPercentile() info.Percentiles {
+	p := info.Percentiles{}
+	p.Mean = uint64(self.mean.Mean)
+	p.Max = self.max
+	p.Ninety = self.samples.Get90Percentile()
+	p.Present = true
+	return p
+}
+
+func NewResource(size int) *Resource {
+	return &Resource{
+		samples: make(uint64Slice, 0, size),
+		mean:    Mean{count: 0, Mean: 0},
+	}
+}
+
+// Return aggregated percentiles from the provided percentile samples.
+func GetDerivedPercentiles(stats []*info.Usage) info.Usage {
+	cpu := NewResource(len(stats))
+	memory := NewResource(len(stats))
+	for _, stat := range stats {
+		cpu.Add(stat.Cpu)
+		memory.Add(stat.Memory)
+	}
+	usage := info.Usage{}
+	usage.Cpu = cpu.GetPercentile()
+	usage.Memory = memory.GetPercentile()
+	return usage
+}
+
+// Calculate part of a minute this sample set represent.
+func getPercentComplete(stats []*secondSample) (percent int32) {
+	numSamples := len(stats)
+	if numSamples > 1 {
+		percent = 100
+		timeRange := stats[numSamples-1].Timestamp.UnixNano() - stats[0].Timestamp.UnixNano()
+		// allow some slack
+		if timeRange < 58*secondsToNanoSeconds {
+			percent = int32((timeRange * 100) / 60 * secondsToNanoSeconds)
+		}
+	}
+	return
+}
+
+// Returns a percentile sample for a minute by aggregating seconds samples.
+func GetMinutePercentiles(stats []*secondSample) info.Usage {
+	lastSample := secondSample{}
+	cpu := NewResource(len(stats))
+	memory := NewResource(len(stats))
+	for _, stat := range stats {
+		var elapsed int64
+		if !lastSample.Timestamp.IsZero() {
+			elapsed = stat.Timestamp.UnixNano() - lastSample.Timestamp.UnixNano()
```

Comment:
nit: can do this with Sub() and time.Duration

if t1.Sub(t1) < 10 * time.Nanosecond {
}